### PR TITLE
More user info

### PR DESCRIPTION
**Pull request by Prashantneu8848**
This PR is build on top of PR #15 and entails the following changes:
1. UserInfoModalBox.js -> Modal Box for the user to enter their information.
2. LoginInterface.java -> Interface which is implemented by LoginService.java
3. Added fields on UserInfo.java for first name, last name, user name and favorite store name
4. LoginService.java -> abstracted logics of LoginServlet
5. Added component on TopNavBar for UserInfoModalBox
6. Changed generic name modal to uploadModal
7. Related tests for these changes.

UserInfoModal -> https://drive.google.com/file/d/1qTy21SsqKe5N_g957RbAjYqPdfgDXyQO/view?usp=sharing
Display Name before User Info -> https://drive.google.com/file/d/1i95UNm0g_08MnWBmFERZQOttKxupRaow/view?usp=sharing
Display Name after User Info -> https://drive.google.com/file/d/1jtgLxM5emWMjcsanmW3RTA__7XqwwWfy/view?usp=sharing

### CODE REVIEW

File: src/test/java/com/google/edith/LoginServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    testHelper.tearDown();
+  }
+
+  @Mock HttpServletRequest request;
+
+  // Checks the login status of the user.
+  @Test
+  public void checkUserLoggedIn_returnsUserLoggedInStatus() {
+    assertTrue(loginService.checkUserLoggedIn());
+    assertTrue(userService.isUserLoggedIn());
```
[the remark this answers is not in the document]
Comment (Prashantneu8848) [reply src/test/java/com/google/edith/LoginServiceTest.java:76]: Thanks for your suggestions.
I think having multiple assertions in the same test creates a problem when the test fails and they are testing different behaviors.
Here, either one of them can only be used as both of them tests the same behavior. But ```assertTrue(userService.isUserLoggedIn());``` might be better as it is from https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/users/UserService

File: edith-step/src/components/FileUploadModalBox.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import Button from 'react-bootstrap/Button';
+import Form from 'react-bootstrap/Form';
+import Modal from 'react-bootstrap/Modal';
+import PropTypes from 'prop-types';
+
+/**
+ * Modal Box where user can select a file to upload.
+ */
+class FileUploadModalBox extends React.Component {
+  /**
```
Comment (malachibre) [inline edith-step/src/components/FileUploadModalBox.js:11]: Is there supposed to be a space between the class declaration and the comment?

File: src/main/java/com/google/edith/interfaces/LoginInterface.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.edith.interfaces;
+
+import javax.servlet.http.HttpServletRequest;
+
+/** Operations for handling user authentication. */
+public interface LoginInterface {
+  public String createJsonOfUserInfo();
```
Comment (malachibre) [inline src/main/java/com/google/edith/interfaces/LoginInterface.java:21]: Nit: You can remove the public keyword for methods in an interface.

File: edith-step/src/components/FileUploadModalBox.js (hunk 2 above)
Comment (Prashantneu8848) [reply edith-step/src/components/FileUploadModalBox.js:11]: hmm, i guess the general form is to be supposed to be that way ?
https://google.github.io/styleguide/jsguide.html#jsdoc-general-form

File: src/main/java/com/google/edith/interfaces/LoginInterface.java (hunk 3 above)
Comment (Prashantneu8848) [reply src/main/java/com/google/edith/interfaces/LoginInterface.java]: Thanks.
Changes made.

File: src/test/java/com/google/edith/BlobstoreUrlServletTest.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  /** Checks if the url contains required path. */
+  @Test
+  public void doGet_containsRightUrl() throws IOException {
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+    blobstoreUrlServlet.doGet(request, response);
+    verify(response, atLeast(1)).getWriter();
+    assertTrue(stringWriter.toString().contains("/_ah/upload/"));
+  }
```
Comment (sehtia-raghav) [inline src/test/java/com/google/edith/BlobstoreUrlServletTest.java:68]: It's ideal to separate test methods into three logical chunks: arrange, act, and assert. Arrange is where you organize all your input data for your tests, act is the method you're testing, and assert is assert lol.

File: src/test/java/com/google/edith/LoginServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+  private LoginService loginService;
+  private final UserService userService = UserServiceFactory.getUserService();
+  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+    testHelper.setUp();
+    loginService = new LoginService(userService, datastore);
```
Comment (sehtia-raghav) [inline src/test/java/com/google/edith/LoginServiceTest.java:62]: If this is being setup before every test can we not initialize it directly in line 54? you'd probably have to move 54 after 56

File: src/test/java/com/google/edith/LoginServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+    assertTrue(userInfo.contains("firstName"));
+    assertTrue(userInfo.contains("lastName"));
+    assertTrue(userInfo.contains("favoriteStore"));
+    assertTrue(userInfo.contains("email"));
+    assertTrue(userInfo.contains("logOutUrl"));
+  }
+
+  // Creates a stub user info for testing.
+  private void retrieveStubUserInfo(HttpServletRequest request) {
```
Comment (sehtia-raghav) [inline src/test/java/com/google/edith/LoginServiceTest.java:117]: nit: I think maybe "mockUserInfo" or "mockUserInfoFromRequest" or something along those lines that indicates user info in the request is being mocked might be a more descriptive name .

File: src/test/java/com/google/edith/LoginServletTest.java [hunk 7]
```diff
@@ -41,75 +33,84 @@
...
-      new LocalServiceTestHelper(new LocalUserServiceTestConfig()).setEnvIsLoggedIn(false);
-
-  private final UserService userService = UserServiceFactory.getUserService();
   private LoginServlet loginServlet;
 
   @Mock HttpServletRequest request;
 
   @Mock HttpServletResponse response;
 
+  @Mock LoginInterface loginImpl;
```
Comment (sehtia-raghav) [inline src/test/java/com/google/edith/LoginServletTest.java:43]: nit: loginService or similar to avoid abbreviations

Comment (Prashantneu8848) [reply src/test/java/com/google/edith/LoginServletTest.java:43]: Changed it to ```loginImplementation```

File: src/test/java/com/google/edith/LoginServiceTest.java (hunk 6 above)
Comment (Prashantneu8848) [reply src/test/java/com/google/edith/LoginServiceTest.java:117]: Thanks for the suggestion. Changed it to ```mockUserInfoFromRequest```

File: src/test/java/com/google/edith/LoginServiceTest.java (hunk 5 above)
Comment (Prashantneu8848) [reply src/test/java/com/google/edith/LoginServiceTest.java:62]: Yes, we can. Thanks for the suggestion.

File: src/test/java/com/google/edith/BlobstoreUrlServletTest.java (hunk 4 above)
Comment (Prashantneu8848) [reply src/test/java/com/google/edith/BlobstoreUrlServletTest.java:68]: Thanks for the suggestion.
Changes made.

File: src/main/java/com/google/edith/servlets/LoginServlet.java [hunk 8]
```diff
@@ -31,44 +31,37 @@
 @WebServlet("/login")
 public final class LoginServlet extends HttpServlet {
 
+  private LoginInterface loginImpl;
```
[the remark this answers is not in the document]
Comment (Prashantneu8848) [reply src/main/java/com/google/edith/servlets/LoginServlet.java:34]: thanks !

File: src/main/java/com/google/edith/servlets/LoginServlet.java [hunk 9]
```diff
@@ -31,44 +31,37 @@
...
     } else {
-      String loginUrl = userService.createLoginURL("/");
+      String loginUrl = loginImpl.createLoginUrl("/");
       response.sendRedirect(loginUrl);
     }
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // User who are not logged in can not upload their info. This is handled in FE too.
```
[the remark this answers is not in the document]
Comment (Prashantneu8848) [reply src/main/java/com/google/edith/servlets/LoginServlet.java:61]: Yes, thanks for catching that. Changed it to users who are not ...

File: src/main/java/com/google/edith/servlets/LoginServlet.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Prashantneu8848) [reply src/main/java/com/google/edith/servlets/LoginServlet.java:34]: Changed it to ```loginImplementation```

File: src/main/java/com/google/edith/services/ReceiptFileHandlerService.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+    Map<String, List<FileInfo>> fileInfos = blobstoreService.getFileInfos(request);
+    List<FileInfo> uploadedFile = fileInfos.get(formInputElementName);
+    return uploadedFile == null ? Collections.emptyList() : uploadedFile;
+  }
+
+  /** Returns a BlobKey that points to the uploaded file. */
+  public BlobKey getBlobKey(List<FileInfo> fileKeys) {
+    if (fileKeys.isEmpty()) {
+      throw new IllegalStateException();
```
[the remark this answers is not in the document]
Comment (Prashantneu8848) [reply src/main/java/com/google/edith/services/ReceiptFileHandlerService.java:57]: most definitely.

File: src/main/java/com/google/edith/services/LoginService.java [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+  }
+
+  /**
+   * Returns the UserInfo entity with user id. Given id is not of UserInfo kind but a field of that
+   * kind.
+   *
+   * @param id - id of the user who is logged in.
+   */
+  private Optional<Entity> getUserInfoEntity(String id) {
+
```
[the remark this answers is not in the document]
Comment (Prashantneu8848) [reply src/main/java/com/google/edith/services/LoginService.java:113]: most definitely

File: src/main/java/com/google/edith/services/LoginService.java [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    this.userService = userService;
+    this.datastore = datastore;
+  }
+
+  /**
+   * Converts UserInfo object to JSON string.
+   *
+   * @return String - JSON string.
+   */
+  public String createJsonOfUserInfo() {
```
[the remark this answers is not in the document]
Comment (Prashantneu8848) [reply src/main/java/com/google/edith/services/LoginService.java:46]: Changes made.

File: edith-step/src/components/TopNavbar.js [hunk 13]
```diff
@@ -17,7 +19,26 @@ class TopNavbar extends React.Component {
...
+      uploadModalBoxShow: false,
+      userInfoModalBoxShow: false,
     };
+    this.handleUploadModalClose = this.handleUploadModalClose.bind(this);
+    this.handleUserInfoModalBoxClose =
+        this.handleUserInfoModalBoxClose.bind(this);
+  }
+
+  /**
+   * Call back function to close the file upload modal box.
```
[the remark this answers is not in the document]
Comment (Prashantneu8848) [reply edith-step/src/components/TopNavbar.js:31]: Changes made.

**Pull request merged**